File: 10003 - Cutting Sticks/main.py
from sys import stdin

# TODO: Too slow for now see main.c

def readstick():
    """Read one stick problem form stdin"""
    length = int(stdin.readline())
    if not length:
        return None, None

    ncuts = int(stdin.readline())

    cuts = list(map(int, stdin.readline().split()))

    return length, cuts


def cost(s, e, cuts, mem):
    """Minimal cost of cuts from s to e
    Arguments: start, end, cuts, mem"""
    if (s,e) in mem:
        return mem[s, e]

    # Didn't have the cost stored, compute and store
    best = min(cost(s, i, cuts, mem)+cost(i, e, cuts, mem) for i in range(s+1, e))
    mem[s, e] = best + cuts[e]-cuts[s]
    return mem[s,e]
    

def min_cost(length, cuts):
    """Find minimum cost using recursion with memoization""" 
    cuts = [0] + cuts + [length]

    # Cost table
    mem = {}
    for c in range(len(cuts)-1):
        mem[c, c+1] = 0

    return cost(0, len(cuts)-1, cuts, mem)
       

if __name__ == '__main__':
    
    length, cuts = readstick()
    while length:
        print('The minimum cutting is {}.'.format(min_cost(length, cuts)))
        length, cuts = readstick()

File: 10003 - Cutting Sticks/main.c
#include <stdio.h>
#include <stdlib.h>
#include <limits.h>
#include <string.h>

#define min(a,b) \
	({ __typeof__ (a) _a = (a); \
	 __typeof__ (b) _b = (b); \
	 _a < _b ? _a : _b; })

#define MAX_CUTS 50
#define MAX_LENGTH 1000

int mem[MAX_CUTS+2][MAX_CUTS+2];
int cuts[MAX_CUTS+2];
int ncuts = 0;

int load_stick(void){
	/* Load next stick problem */
	int length, i;
	
	scanf("%d", &length);
	if (length==0)
		return 0;

	scanf("%d", &ncuts);
	for(i=0; i<ncuts; i++)
		scanf("%d", &cuts[i+1]);
	cuts[ncuts+1]=length;
	cuts[0]=0;
	ncuts+=2;
	return 1;
}



int cost(s, e) {
	int i, best_cost=INT_MAX;

	if (mem[s][e] >= 0)
		return mem[s][e];

	for(i=s+1; i<e; i++)
		best_cost = min(cost(s,i)+cost(i,e), best_cost);

    return mem[s][e] = best_cost + cuts[e] - cuts[s];
}

int min_cost(void) {
	short i;

	/* Initialize array from last stick */
	memset(mem, -1, sizeof(mem));	
	for(i=0; i<ncuts; i++) 
	   mem[i][i+1] = 0;

	/* Find minimum cost */
	return cost(0, ncuts-1);
}



int main(int argc, char *argv[]) {
	while(load_stick())
		printf("The minimum cutting is %d.\n", min_cost());
}
